### PR TITLE
Fix to Customize controls 

### PR DESCRIPTION
**Pull request by Olzaq**

### CODE REVIEW

File: src/core/Config.h [hunk 1]
```diff
@@ -203,7 +212,7 @@ class Config {
 	
 public:
 	
-	void setActionKey(ControlAction action, int index, InputKeyId key);
+	void setActionKey(const ControlAction action, uint32_t index, InputKeyId newKey);
```
Comment (dscharrer) [inline src/core/Config.h]: There is no reason to add const to a non reference, non-pointer parameter in a function prototype - it has literally no meaning.

File: src/core/Config.cpp [hunk 2]
```diff
@@ -337,44 +337,55 @@ void Config::setDefaultActionKeys() {
...
-			}
-		}
-		
-	}
+	moveOldKeyToOtherIndexIfEmpty(otherIndex, oldKey, action);
+	clearOtherIndexIfDuplicate(otherIndex, newKey, action);
+	removeDuplicateKeyAssignments(actionId, newKey, oldKey);
+}
+
+void Config::moveOldKeyToOtherIndexIfEmpty(uint32_t otherIndex,
```
Comment (dscharrer) [inline src/core/Config.cpp]: This commit seems to have lots of unrelated changes. Also I'm not sure if splitting out these tiny functions makes things more readable - their code is a lot easier to understand than the function names.

File: src/core/Config.cpp [hunk 3]
```diff
@@ -337,44 +337,55 @@ void Config::setDefaultActionKeys() {
 	}
 }
 
-void Config::setActionKey(ControlAction actionId, int index, InputKeyId key) {
+void Config::setActionKey(const ControlAction actionId, uint32_t index, InputKeyId newKey) {
```
Comment (dscharrer) [inline src/core/Config.cpp]: Why make the index a uint32_t?

File: src/core/Config.cpp [hunk 4]
```diff
@@ -348,7 +348,7 @@ void Config::setActionKey(const ControlAction actionId, uint32_t index, InputKey
 	InputKeyId oldKey = action.key[index];
 	action.key[index] = newKey;
 	
-	uint32_t otherIndex = ALT_ACTION - index;
+	uint32_t otherIndex = ALT_KEY - index;
```
Comment (dscharrer) [inline src/core/Config.cpp]: This name is very misleading -  "alt key" already has a well known meaning.

File: src/core/Config.h [hunk 5]
```diff
@@ -90,17 +90,17 @@ enum UIScaleFilter {
 namespace {
 
 const uint32_t NUM_OF_KEYS_PER_ACTION = 2;
-const uint32_t DEFAULT_ACTION = 0;
-const uint32_t ALT_ACTION = 1;
+const uint32_t DEFAULT_KEY = 0;
```
Comment (dscharrer) [inline src/core/Config.h]: DEFAULT_KEY is not very descriptive.

File: src/core/Config.h [hunk 6]
```diff
@@ -90,17 +90,17 @@ enum UIScaleFilter {
 namespace {
```
Comment (dscharrer) [inline src/core/Config.h]: No anonymous namespace in headers! (Also they might not do what you expect in source files due to unity builds.)

File: src/core/Config.h [hunk 7]
```diff
@@ -86,15 +86,24 @@ enum UIScaleFilter {
 	UIFilterBilinear = 1
 };
 
+
+namespace {
+
+const uint32_t NUM_OF_KEYS_PER_ACTION = 2;
+const uint32_t DEFAULT_ACTION = 0;
```
Comment (dscharrer) [inline src/core/Config.h]: There is nothing "default" about the first action binding. Also introducing names and then renaming then in the same commit series is pointless.

File: src/core/Config.cpp (hunk 3 above)
Comment (Olzaq) [inline src/core/Config.cpp]: It makes the value range check easier as >= can be used to check the whole meaningful range.

File: src/core/Config.cpp (hunk 2 above)
Comment (Olzaq) [inline src/core/Config.cpp]: Is using small functions like these something you prefer not to be used in the project? Even in a code like menus, configs etc. as this code is not performance critical (and compiler optimize them away anyway). Writing functions instead of comments makes a code generally (and ignoring this case now) more readable I think.

File: src/core/Config.cpp (hunk 3 above)
Comment (dscharrer) [inline src/core/Config.cpp]: Sure, my comment was more about it being 32-bit. While `int` is also 32-bit in practice it does not mean that it *mus* be 32-bit.

Like, Eli2 commented on irc, size_t (or a signed int) should be used for indices. Smaller fixed sizes for indices only make sense as a performance optimization (not important here) and in this case it would only make things worse (see the video Eli2 liked in irc if you are interested in why).

Comment (Olzaq) [inline src/core/Config.cpp]: true, just checked the vid.

File: src/core/Config.cpp (hunk 2 above)
Comment (dscharrer) [inline src/core/Config.cpp]: It really depends on how small they and on the name & arguments. The point is to make the code as readable as possible not to have some absolute rules about using functions instead of comments.

Some concrete points:
* For the first two functions, the code is so short it its obvious what it does and doesn't need a comment. Moving it into a function only requires the reader to jump around in the file to follow the code flow and understand the overall bigger function.
* Your extracted function names describe what the functions are used for instead of what they do: `moveOldKeyToOtherIndexIfEmpty` does not care that the key is old, `moveOldKeyToOtherIndexIfEmpty` and `clearOtherIndexIfDuplicate` don't care about the index not being the primary one - they only deal with one index, and `removeDuplicateKeyAssignments` does not care about duplicates at all - it simply replaces the first use of a key with a supplied value and all other uses (there should be none if the bindings already are consistent!) with `UNASSIGNED_KEY`.
* The names `moveOldKeyToOtherIndexIfEmpty` and `clearOtherIndexIfDuplicate` do not clearly describe what is checked for being empty/duplicate. The key or index aren't empty, the binding slot is. The index isn't duplicated, the key binding is.
* Having conditions like "if", "except", etc. in your function names is a good indicator that your abstraction is not at the right level.

A *good* way to extract a function here could be:
```cpp
/*!
 * \brief Bind a new key to the action
 *
 * \param index Index of the slot to bind the key to.
 *
 * \return the previously bound key if it could not be moved to the other slow slot,
 *         or \ref UNASSIGNED_KEY otherwise
 */
InputKeyId ActionKey::changeBinding(int index, InputKeyId newKey) {
	int otherIndex = 1 - index;
	InputKeyId oldKey = key[index];
	key[index] = newKey;
	if(key[otherIndex] == UNASSIGNED_KEY) {
		key[otherIndex] = oldKey;
		oldKey = UNASSIGNED_KEY;
	}
	if(key[otherIndex] == newKey) {
		key[otherIndex] = UNASSIGNED_KEY;
	}
	return oldKey;
}
```
Possibly made private or not as a member function though to discourage calling it directly.
That way:
-  The concerns are neatly separated: The extracted function only cares about consistency in one `ActionKey` while the larger `Config::setActionKey` does not need to care about there being multiple bindings per action.
- The function conceptually only modifies one thing (the ActionKey) as opposed to two things in your `moveOldKeyToOtherIndexIfEmpty` (the ActionKey binding at otherIndex and the oldKey value).
- Returning the old value when changing something is a common idiom. The fact that there are two slots and you are only returning something if both slots are full is an implementation detail.
- Using it as `InputKeyId removedKey = action.changeBinding(index, key)` would neatly document what you are doing.

`removeDuplicateKeyAssignments` has a bit more reason for existing as a separate function. However, the code is still not very long, `Config::setActionKey` does not contain much else after `ActionKey::changeBinding` has been extracted, and "replace all uses of this key except for those in this action" does not make for a nice function name. An argument *for* extracting it could be that there is potential to reuse it to sanitize bindings while loading the config.

Hope this helps understand my reasoning and I hope I'm not coming off as too discouraging ;-)

**Comment by dscharrer** [comment]
Superseded by pull request #218.
